agent/runner/runner_test.go

```diff
@@ -121,28 +117,127 @@ func TestDefaultCapacityLimit(t *testing.T) {
 	t.Parallel()
 
 	// Use default capacity
-	cr := New(0)
+	cr := New(0, 0)
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	go cr.Run(ctx)
 
-	totalJobs := 2 * defaultCapacity
+	totalJobs := 2 * defaultTotalCapacity
 	for i := 0; i < totalJobs; i++ {
 		require.NoError(t, cr.StartJob(testJob{id: fmt.Sprintf("test-%d", i), timeout: time.Second}))
 	}
 
-	// Let first jobs start
+	// Let jobs to start
 	time.Sleep(500 * time.Millisecond)
 
+	var running int
 	for i := 0; i < totalJobs; i++ {
 		// Check that running jobs amount is not exceeded default capacity.
-		assert.Equal(t, i < defaultCapacity, cr.IsRunning(fmt.Sprintf("test-%d", i)))
+		if cr.IsRunning(fmt.Sprintf("test-%d", i)) {
+			running++
+		}
 	}
+
+	assert.Equal(t, defaultTotalCapacity, running)
+}
+
+func TestPerDBInstanceLimit(t *testing.T) {
+	t.Parallel()
+
+	cr := New(10, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go cr.Run(ctx)
+
+	j1db1 := testJob{id: "test-1", timeout: time.Second, dsn: "postgresql://db1"}
+	j2db1 := testJob{id: "test-2", timeout: time.Second, dsn: "postgresql://db1"}
+	j3db1 := testJob{id: "test-3", timeout: time.Second, dsn: "postgresql://db1"}
+	j1db2 := testJob{id: "test-4", timeout: time.Second, dsn: "postgresql://db2"}
+	j2db2 := testJob{id: "test-5", timeout: time.Second, dsn: "postgresql://db2"}
+	j3db2 := testJob{id: "test-6", timeout: time.Second, dsn: "postgresql://db2"}
+
+	require.NoError(t, cr.StartJob(j1db1))
+	require.NoError(t, cr.StartJob(j1db2))
+
+	// Let jobs to start
+	time.Sleep(200 * time.Millisecond)
+
+	require.NoError(t, cr.StartJob(j2db1))
+	require.NoError(t, cr.StartJob(j2db2))
+	require.NoError(t, cr.StartJob(j3db1))
+	require.NoError(t, cr.StartJob(j3db2))
+
+	// Let rest jobs to reach semaphores
+	time.Sleep(300 * time.Millisecond)
+
+	assert.True(t, cr.IsRunning(j1db1.ID()))
+	assert.True(t, cr.IsRunning(j1db2.ID()))
+	assert.False(t, cr.IsRunning(j2db1.ID()))
+	assert.False(t, cr.IsRunning(j2db2.ID()))
+	assert.False(t, cr.IsRunning(j3db1.ID()))
+	assert.False(t, cr.IsRunning(j3db2.ID()))
+
+	// Over time all jobs are terminated
+	time.Sleep(time.Second)
+
+	assert.False(t, cr.IsRunning(j1db1.ID()))
+	assert.False(t, cr.IsRunning(j1db2.ID()))
+	assert.False(t, cr.IsRunning(j2db1.ID()))
+	assert.False(t, cr.IsRunning(j2db2.ID()))
+	assert.False(t, cr.IsRunning(j3db1.ID()))
+	assert.False(t, cr.IsRunning(j3db2.ID()))
```

Comment:
does that mean that timeout timer starts from the moment pmm-agent receives the message. previously behaviour was different and cancel timer was starting after acquiring the global semaphore

Reply:
Yes, and it looks like before it wasn't correct. It seems that from user perspective timeout should be like `total time allocated for execution`.

Reply:
as far as I see user doesn't provide this timeout, so it's our internal mechanism to not let action to work forever.

Reply:
By `user` I mean `PMM Server` or anything else that is interested in such action.